common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java

```diff
@@ -374,6 +379,45 @@ public int removeBlocks(String appId, String execId, String[] blockIds) {
       .collect(Collectors.toMap(Pair::getKey, Pair::getValue));
   }
 
+  /**
+   * Diagnose the possible cause of the shuffle data corruption by verify the shuffle checksums
+   */
+  public Cause diagnoseShuffleBlockCorruption(
+      String appId,
+      String execId,
+      int shuffleId,
+      long mapId,
+      int reduceId,
+      long checksumByReader) {
+    ExecutorShuffleInfo executor = executors.get(new AppExecId(appId, execId));
+    String fileName = "shuffle_" + shuffleId + "_" + mapId + "_0.checksum";
+    try {
+      // This's consistent with `IndexShuffleBlockResolver.getChecksumFile`.
+      // We firstly use `fileName` to get the location of the checksum file.
+      // Then, we use `Files.newDirectoryStream` to list all the files under the same directory
+      // with `probeFile`. Since there's only one single checksum file for a certain map task,
+      // so it's supposed to return one matched file too.
```

Comment:
can we let the client side pass the checksum algorithm?

Reply:
Added.